01-architecture.md

```diff
@@ -0,0 +1,222 @@
+
+The FVM aims to (a) support a multitude of programming models for actors and (b) facilitate the onboarding of smart contracts and programs written for other environments, so they can leverage the storage capabilities of the Filecoin network.
+
+Its architecture is inspired by the [hypervisor](https://en.wikipedia.org/wiki/Hypervisor) concept from the world of Virtual Machines. The abstraction is not translated directly, though.
+
+## Native VM
+
+The native runtime will be based on WebAssembly (WASM), with potential usage of a limited set of WebAssembly System Interface (WASI) APIs.
+
+## Built-in system actors
+
+The existing built-in system actors will be compiled to WASM and will run entirely in WASM space, thus preventing any associated FFI barrier costs.
+
+For the sake of analogy, built-in system actors in Filecoin receive comparable treatment to "precompiled contracts" in other platforms. They are specified by the protocol, implicitly deployed, addressed at fixed locations/mailboxes, and they evolve through system upgrades.
+
+**Canonical system actors**
+
+This transition to WASM opens up the opportunity to converge on a single codebase of actors for all Filecoin implementations.
+
+In the past, each team would re-implement actors in their language (although some relied on FFI with Go actors, a slower approach). With FVM, a single codebase can be compiled to WASM (its new portable executable form) and be adopted across all implementations.
+
+We acknowledge this strategy has tradeoffs, but we won't elaborate on them here.
+
+## Actor sandbox
+
+The actor sandbox is the tightly constrained and instrumented environment in which Filecoin actor code runs.
+
+The actor sandbox must be formally specified, and potentially versioned. 
+
+User-deployed actors may have to specify the sandbox version they expect, as this facilitates sandbox evolution while preserving backwards compatibility in an obvious manner.
+
+Sandbox responsibilities:
+
+1. To inject/bind/link implicit module imports that provide capabilities such as state access, chain access, syscalls, IPLD codec processing, and more.
+2. To provide a toolbox of common functions and types, such as BigInt arithmetic, parameter unpacking, method dispatch tables, etc.
+3. To manage memory usage and halt execution when it exceeds the limits.
+4. To load input call parameters and handle return values.
+5. To intercept cross contract calls and delegate to the outer environment.
+6. To perform gas accounting by instrumenting bytecode.
+7. To halt execution when available gas is exhausted.
+
+(1) and (2) together form the basis of the **FVM stdlib**: the set of APIs that are universally available for actors to call inside the actor sandbox.
+
+One key constraint is that actor code size/footprint should be kept as low as possible to avoid state bloat and gas costs. (2) is important, as it precludes repetitive/common code from being statically linked to user-deployed actors.
```

Comment:
The second part isn't _entirely_ true. Given that we deploy once and init many times, initializing new actors should be _fairly_ cheap even if the initial deploy is a bit expensive.

However, this brings up an interesting point: updates. It would be _wonderful_ to be able to "patch" an actor without having to re-upload all of the old code.